Libsystem: remove broken `postFetch` step in `darling.src`

runeksvendsen opened the pull request:
## Description of changes

Fixes #297765

The `fetchzip` `postFetch` step in `darling.src` in `Libsystem` is broken because it references `$downloadedFile`, which does not exist because `fetchzip` moves `$downloadedFile` to `$renamed` before the `postFetch` step is executed.

The brokenness of this step did not become apparent until commit 5a6a31e54ddaf59aa8e1c5e7e9b02623d4881c34 was merged, since before this commit the `fetchzip` `postFetch` step wasn't executed at all.

Furthermore, the tarball referenced by `darling.src` no longer contains both `src/opendirectory` and `src/OpenDirectory`, so the `postFetch` step isn't needed anyway.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
  - [x] `sandbox = false`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wegank commented:
>  the tarball referenced by darling.src no longer contains both `src/opendirectory` and `src/OpenDirectory`

Are you sure?

wegank reviewed (changes requested):
>  the tarball referenced by darling.src no longer contains both `src/opendirectory` and `src/OpenDirectory`

Are you sure?

runeksvendsen commented:
> > the tarball referenced by darling.src no longer contains both `src/opendirectory` and `src/OpenDirectory`
> 
> Are you sure?



> > the tarball referenced by darling.src no longer contains both `src/opendirectory` and `src/OpenDirectory`
> 
> Are you sure?

Oops, you're right. They both exist. However it's not an issue because `darling.src` is inlined into `Libsystem` and `Libsystem` neither uses `src/OpenDirectory` nor its lowercase counterpart. However, I don't mind keeping the `postFetch` step and replacing `$downloadedFile` with `$renamed` if you prefer this.

wegank commented:
The files should already be extracted before `postFetch` in `fetchzip`, so no need to `tar -xzf $renamed` again. I think something like the following would be sufficient.

```nix
postFetch = ''
  rm -rf $out/src/{OpenDirectory,opendirectory}
'';
```

runeksvendsen commented:
> The files should already be extracted before `postFetch` in `fetchzip`, so no need to `tar -xzf $renamed` again. I think something like the following would be sufficient.
> 
> ```nix
> postFetch = ''
>   rm -rf $out/src/{OpenDirectory,opendirectory}
> '';
> ```

@wegank  That's one option. Alternatively, we could keep just the part that merges the two directories. Something like:

```nix
postFetch = ''
  cd $out

  # The archive contains both `src/opendirectory` and `src/OpenDirectory`.
  # If `src/opendirectory` and `src/OpenDirectory` refer to different
  # things, then combine them into `src/OpenDirectory` to match the result
  # on case-insensitive filesystems.
  if [ "$(stat -c %i src/opendirectory)" != "$(stat -c %i src/OpenDirectory)" ]; then
    mv src/opendirectory/* src/OpenDirectory/
    rmdir src/opendirectory
  fi
'';
```

wegank commented:
There may also be a hash mismatch if, on a case-insensitive system, `unpackFile` is not deterministic between `$out/src/opendirectory` and `$out/src/OpenDirectory`.

wegank commented:
Also, the number of new rebuilds is too large, please [rebase onto staging](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging).

runeksvendsen commented:
> The files should already be extracted before `postFetch` in `fetchzip`, so no need to `tar -xzf $renamed` again. I think something like the following would be sufficient.
> 
> ```nix
> postFetch = ''
>   rm -rf $out/src/{OpenDirectory,opendirectory}
> '';
> ```

> There may also be a hash mismatch if, on a case-insensitive system, `unpackFile` is not deterministic between `$out/src/opendirectory` and `$out/src/OpenDirectory`. Keeping just the part that merges the two directories does not prevent this.

Good point on the hash mismatch depending on whether the file system is case sensitive or not.

Question: when are the files hashed? I presume the file specified by `url` is fetched, unpacked and then hashed (without running any step in-between). But I can't tell from browsing the code, since I can't find the code that `unpackFile` runs. I end up here https://github.com/NixOS/nixpkgs/blob/0b672c78ab67e9611f66714a5016ce37bd19a4d4/pkgs/stdenv/generic/setup.sh#L881 where I can't tell where `unpackCmd` is defined.

wegank commented:
Tarballs are handled in `_defaultUnpack` in `pkgs/stdenv/generic/setup.sh`, the same file you've mentioned.

runeksvendsen commented:
@wegank thank you. Given the files inside `OpenDirectory`/`opendirectory` aren't used currently, I think your suggestion to just remove these two directories is the way to go since it avoids the potential issue with a hash mismatch depending on the case sensitivity of the filesystem. It's possible this isn't an issue at all, but I don't want to spend time investigating it when the files inside these two directories aren't needed anyway.

I will force push a commit to this PR branch which uses this approach with commit message explaining the reasoning.

runeksvendsen committed:
`Libsystem`: fix broken `postFetch` step in  `darling.src`

Fixes #297765

The `postFetch` step was broken in several ways:
1. It attempts to unpack the archive, but this has already been done before reaching this step.
2. It unpacks the file `$downloadedFile`, but this file no longer exists because it's been moved to `$renamed`.

Additionally, since (1) neither directory `src/opendirectory` nor `src/OpenDirectory` are used for anything, and (2) it's not clear whether Nix produces different hashes depending on file system case sensitivity, the chosen solution is to just remove both these two directories for now. If later a file from either directory is needed, an investigation needs to be done on how Nix hashes archives with clashing names on case insensitive file systems.

wegank commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix:
```diff
@@ -12,22 +12,11 @@ let
     url = "https://github.com/darlinghq/darling/archive/d2cc5fa748003aaa70ad4180fff0a9a85dc65e9b.tar.gz";
     sha256 = "11b51fw47nl505h63bgx5kqiyhf3glhp1q6jkpb6nqfislnzzkrf";
```
Please also update the hash.

runeksvendsen replied:
Whoops, sorry. Things went a bit fast. Done.

wegank commented:
@ofborg build hello
